src/gluonnlp/data/data_preprocessing_transform.py

```diff
@@ -0,0 +1,421 @@
+                    whitespace_splitter(orig_answer_text.strip()))
+                if actual_text.find(cleaned_answer_text) == -1:
+                    print('Could not find answer: %s vs. %s' %
+                          (actual_text, cleaned_answer_text))
+                    return None
+            else:
+                start_position = -1
+                end_position = -1
+                orig_answer_text = ''
+
+        example = SquadExample(qas_id=qas_id,
+                               question_text=question_text,
+                               doc_tokens=doc_tokens,
+                               example_id=example_id,
+                               orig_answer_text=orig_answer_text,
+                               start_position=start_position,
+                               end_position=end_position,
+                               is_impossible=is_impossible)
+        return example
+
+
+class TokenizeAndPositionAlignTransform:
+    """Tokenize the question and paragraph text and map the origin start/end position
+    to the right position in tokenized text.
+    """
+    def __init__(self, tokenizer, max_query_length, is_training):
+        self._tokenizer = tokenizer
+        self._max_query_length = max_query_length
+        self.is_training = is_training
+
+    def __call__(self, example):
+        # tokenize the query text
+        query_tokens = self._tokenizer(example.question_text)
+        if len(query_tokens) > self._max_query_length:
+            query_tokens = query_tokens[0:self._max_query_length]
+
+        tok_to_orig_index = []
+        orig_to_tok_index = []
+        all_doc_tokens = []
+
+        for (i, token) in enumerate(example.doc_tokens):
+            orig_to_tok_index.append(len(all_doc_tokens))
+            sub_tokens = self._tokenizer(token)
+            tok_to_orig_index += [i] * len(sub_tokens)
+            all_doc_tokens += sub_tokens
+        # tokenize the paragraph text and align the start/end position
+        # to the tokenized sequence position
+        tok_start_position = None
+        tok_end_position = None
+        if self.is_training and example.is_impossible:
+            tok_start_position = -1
+            tok_end_position = -1
+        if self.is_training and not example.is_impossible:
+            tok_start_position = orig_to_tok_index[example.start_position]
+            if example.end_position < len(example.doc_tokens) - 1:
+                tok_end_position = orig_to_tok_index[example.end_position +
+                                                     1] - 1
+            else:
+                tok_end_position = len(all_doc_tokens) - 1
+            (tok_start_position, tok_end_position) = improve_answer_span(
+                all_doc_tokens, tok_start_position, tok_end_position,
+                self._tokenizer, example.orig_answer_text)
+        return tok_start_position, tok_end_position, all_doc_tokens, query_tokens
+
+
+class DocSpanTransform:
```

Comment:
I'm not convinced that having `DocSpan`s to denote windows over a list of tokens is helpful. I think the code is simpler when you materialize each sliding window: Consider a document with 1000 tokens, `doc = [0] * 1000`. We can get the windows explicitly as `window1 = doc[0:100]; window2 = doc[50:150]` etc. Of course you need to calculate the window-specific start and end positions when creating the window. But this should allow removing a lot of the code where you currently have to handle `DocSpan` objects. Instead each window will be completely self-contained after the transform.

As each window is just a list of integers, there is very little overhead in materializing the windows. Lists of integers are highly optimized in Python and we can optimizer more with numpy if needed. I think the introduction of DocSpan here is a sign of premature optimization.